### ISSUE TITLE
dead-lock in SLARTG called by CTGSJA

### ISSUE THREAD

**Issue by winice-test** [Reference-LAPACK/lapack#411]
Hi

When I worked on the GGSVD3 code with some simple case, I found a dead lock in SLARTG which is called by CTGSJA.

The dead-lock occurs between line 161 and 167 when the input argument F equals to INF. But I think the real problem is in CTGSJA line 615-634. There is a check for A1 .NE. ZERO. But when we get GAMMA = B1/A1, we may get an overflow and then calls SLARTG with F=INF at line 623.

Here is a simplified case
```
      PROGRAM tcggsvd3
      IMPLICIT NONE
      COMPLEX*8 A(6,6), B(6,6)
      COMPLEX*8 U(6,6), V(6,6), Q(6,6)
      COMPLEX*8 WORK(1024)
      REAL*4 ALPHA(6), BETA(6)
      REAL*4 RWORK(12)
      INTEGER K,L,IWORK(6), INFO

      A=CMPLX(1.E0, 0.E0)
      B=CMPLX(0.E0, 0.E0)
      B(1,1) = CMPLX(9.E19, 0.E0)
      B(2,2) = CMPLX(9.E18, 0.E0)
      B(3,3) = CMPLX(9.E17, 0.E0)
      B(4,4) = CMPLX(9.E16, 0.E0)
      B(5,5) = CMPLX(9.E15, 0.E0)
      B(6,6) = CMPLX(9.E14, 0.E0)

      CALL CGGSVD3('N','N','N', 6,6,6, K,L, A,6,B,6,
     &             ALPHA,BETA,U,1,V,1,Q,1,WORK,1024,RWORK,IWORK,INFO)
      WRITE(6,*)'INFO = ',INFO
      WRITE(6,*)'ALPHA: ',ALPHA(1:6)
      WRITE(6,*)'BETA:  ',BETA(1:6)

      END PROGRAM tcggsvd3
```

**Comment by thijssteel** [comment]
I think they should both be fixed.

Most importantly, even if an invalid input is supplied, LAPACK should never end up in an infinite loop. So CLARTG, DLARTG,... really should be fixed. I've run into this problem in other places as well. I can make a PR for this after my QZ implementation if finished.

**Comment by martin-frbg** [comment]
#196 is/was a similar thing, I guess it all hinges on knowing a quick, portable and reliable test for INF

**Comment by thijssteel** [comment]
When the input is INF of NAN, the loop does nothing, so we could check whether the values from the last iteration are different from the current ones.

However, the solution you made in that PR seems better to me, just a reasonably large maximum number of iterations

**Comment by langou** [comment]
[ for reference, no action needed ]
Related see http://www.netlib.org/lapack/lawnspdf/lawn148.pdf
where Jim and team provide carefully analyzed implementations of xLARTG.
It includes a MAXCNT to avoid infinite loops (code in Appendices A and B).

**Comment by langou** [comment]
Hi all, 

This issue (#411) is fixed by PR #493. So I am hereby closing the issue now. Thanks @winice-test for the bug report. And thanks @weslleyspereira for this PR #493. 

Note that, while working on fixing this issue,  @weslleyspereira might have found another problem in CTGSJA and is working on a fix. See PR #502. We will continue the work on CTGSJA in PR #502. 

This issue (#411) is fixed. Closing the issue. 
Cheers,
Julien.
